File: modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-02-preparacao-datasets/dataset-preparation-pipeline-java/src/test/java/com/datasetprep/cleaning/BalancingTest.java
package com.datasetprep.cleaning;

import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

class BalancingTest {

    @Test
    void alpha1SemSuavizacaoReproduzADistribuicaoProporcionalOriginal() {
        Map<String, Integer> contagens = Map.of("A", 8, "B", 2);
        var pesos = Balancing.pesosAmostragemPorTemperatura(contagens, 1);
        assertThat(pesos.get("A")).isCloseTo(0.8, org.assertj.core.data.Offset.offset(1e-9));
        assertThat(pesos.get("B")).isCloseTo(0.2, org.assertj.core.data.Offset.offset(1e-9));
    }

    @Test
    void alphaMenorSuavizaADistribuicaoEmDirecaoAUniforme() {
        Map<String, Integer> contagens = Map.of("A", 8, "B", 2);
        var pesosBaixoAlpha = Balancing.pesosAmostragemPorTemperatura(contagens, 0.3);
        var pesosAltoAlpha = Balancing.pesosAmostragemPorTemperatura(contagens, 1);
        assertThat(pesosBaixoAlpha.get("A")).isLessThan(pesosAltoAlpha.get("A"));
        assertThat(pesosBaixoAlpha.get("B")).isGreaterThan(pesosAltoAlpha.get("B"));
    }

    @Test
    void alocacaoSemRestricaoDeCapacidadeSomaExatamenteAoAlvo() {
        Map<String, Double> pesos = Map.of("A", 0.5, "B", 0.3, "C", 0.2);
        var aloc = Balancing.alocarMaiorResto(pesos, 17);
        int soma = aloc.values().stream().mapToInt(Integer::intValue).sum();
        assertThat(soma).isEqualTo(17);
    }

    @Test
    void alocacaoCapacitadaNuncaAlocaMaisQueACapacidadeRealDeNenhumaFonte() {
        Map<String, Integer> contagens = Map.of(
                "Oficina Estrela", 14, "Auto Center Silva", 5, "Funilaria Rio Bonito", 4, "Oficina Nova Alianca", 3);
        for (int alvo : new int[]{26, 22, 20, 18}) {
            var aloc = Balancing.alocarComCapacidade(contagens, Balancing.ALPHA_TEMPERATURA, alvo);
            for (var f : aloc.keySet()) {
                assertThat(aloc.get(f)).as(f).isLessThanOrEqualTo(contagens.get(f));
            }
        }
    }

    @Test
    void alocacaoCapacitadaSomaExatamenteAoAlvoEmTodosOsCasosTestados() {
        Map<String, Integer> contagens = Map.of(
                "Oficina Estrela", 14, "Auto Center Silva", 5, "Funilaria Rio Bonito", 4, "Oficina Nova Alianca", 3);
        for (int alvo : new int[]{26, 22, 20, 18}) {
            var aloc = Balancing.alocarComCapacidade(contagens, Balancing.ALPHA_TEMPERATURA, alvo);
            int soma = aloc.values().stream().mapToInt(Integer::intValue).sum();
            assertThat(soma).isEqualTo(alvo);
        }
    }
}

File: modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-02-preparacao-datasets/dataset-preparation-pipeline-go/cleaning/balancing_test.go
package cleaning

import (
	"math"
	"testing"
)

func TestAlpha1SemSuavizacaoReproduzDistribuicaoProporcional(t *testing.T) {
	contagens := map[string]int{"A": 8, "B": 2}
	pesos := PesosAmostragemPorTemperatura(contagens, 1)
	if math.Abs(pesos["A"]-0.8) > 1e-9 || math.Abs(pesos["B"]-0.2) > 1e-9 {
		t.Fatalf("esperado A=0.8 B=0.2, obtive %+v", pesos)
	}
}

func TestAlphaMenorSuavizaEmDirecaoAUniforme(t *testing.T) {
	contagens := map[string]int{"A": 8, "B": 2}
	baixo := PesosAmostragemPorTemperatura(contagens, 0.3)
	alto := PesosAmostragemPorTemperatura(contagens, 1)
	if !(baixo["A"] < alto["A"]) || !(baixo["B"] > alto["B"]) {
		t.Fatalf("esperado suavizacao em direcao a uniforme, obtive baixo=%+v alto=%+v", baixo, alto)
	}
}

func TestAlocacaoSemRestricaoSomaExatamenteAoAlvo(t *testing.T) {
	pesos := map[string]float64{"A": 0.5, "B": 0.3, "C": 0.2}
	aloc := AlocarMaiorResto(pesos, 17)
	soma := 0
	for _, v := range aloc {
		soma += v
	}
	if soma != 17 {
		t.Fatalf("esperado soma 17, obtive %d", soma)
	}
}

func TestAlocacaoCapacitadaNuncaExcedeCapacidade(t *testing.T) {
	contagens := map[string]int{"Oficina Estrela": 14, "Auto Center Silva": 5, "Funilaria Rio Bonito": 4, "Oficina Nova Alianca": 3}
	for _, alvo := range []int{26, 22, 20, 18} {
		aloc := AlocarComCapacidade(contagens, AlphaTemperatura, alvo)
		for f, n := range aloc {
			if n > contagens[f] {
				t.Errorf("alvo=%d: %s alocado %d > capacidade %d", alvo, f, n, contagens[f])
			}
		}
	}
}

func TestAlocacaoCapacitadaSomaExatamenteAoAlvo(t *testing.T) {
	contagens := map[string]int{"Oficina Estrela": 14, "Auto Center Silva": 5, "Funilaria Rio Bonito": 4, "Oficina Nova Alianca": 3}
	for _, alvo := range []int{26, 22, 20, 18} {
		aloc := AlocarComCapacidade(contagens, AlphaTemperatura, alvo)
		soma := 0
		for _, n := range aloc {
			soma += n
		}
		if soma != alvo {
			t.Errorf("alvo=%d: soma obtida %d", alvo, soma)
		}
	}
}
